commands/new_function_test.go

```diff
@@ -5,37 +5,334 @@ package commands
+		funcName:    "new-test-append-invalid-2",
+		funcLang:    "python3",
+		file:        "invalid2.yml",
+		expectedMsg: InvalidYAMLMsg,
+	},
+	{
+		title:       "new_append_invalid_3",
+		funcName:    "new-test-append-invalid-3",
+		funcLang:    "python",
+		file:        "invalid3.yml",
+		expectedMsg: InvalidYAMLMsg,
+	},
+	{
+		title:       "new_append_invalid_4",
+		funcName:    "new-test-append-invalid-4",
+		funcLang:    "python",
+		file:        "invalid4.yml",
+		expectedMsg: InvalidYAMLMsg,
+	},
+	{
+		title:       "invalid_4",
+		funcName:    "new-test-invalid-1",
+		funcLang:    "dockerfilee",
+		file:        "",
+		expectedMsg: LangNotExistsOutput,
+	},
+}
+
+func parseYAMLFileForNewTest(t *testing.T, fileName string) *stack.Services {
+	parsedService, err := stack.ParseYAMLFile(fileName, "", "")
+	if err != nil {
+		t.Fatalf("Error encountered in file \"%s\": %v", fileName, err)
+	}
+	return parsedService
+}
+
+func runNewFunctionTest(t *testing.T, nft NewFunctionTest) {
+	funcName := nft.funcName
+	funcLang := nft.funcLang
 
-	// Cleanup the created directory
+	cmdParameters := []string{
+		"new",
+		funcName,
+		"--lang=" + funcLang,
+	}
+
+	funcYAML := funcName + ".yml"
+
+	// After the test is complete, reset the test YAML file (that has been
+	// appended to), and clean up the newly created directory.
 	defer func() {
 		os.RemoveAll(funcName)
-		os.Remove(funcName + ".yml")
-		os.Remove("template")
+		os.Remove(funcYAML)
 	}()
 
 	stdOut := test.CaptureStdout(func() {
-		faasCmd.SetArgs([]string{
-			"new",
-			funcName,
-			"--lang=python",
-		})
+		faasCmd.SetArgs(cmdParameters)
+		faasCmd.Execute()
+	})
+
+	// Validate new function output
+	if found, err := regexp.MatchString(nft.expectedMsg, stdOut); err != nil || !found {
+		t.Fatalf("Output is not as expected:\n%s", stdOut)
+	}
+
+	if nft.expectedMsg == SuccessMsg {
+
+		// Make sure that the folder and file was created:
+		if _, err := os.Stat("./" + funcName); os.IsNotExist(err) {
+			t.Fatalf("%s/ directory was not created", funcName)
+		}
+
+		if _, err := os.Stat(funcYAML); os.IsNotExist(err) {
+			t.Fatalf("\"%s\" yaml file was not created", funcYAML)
+		}
+
+		// Make sure that the information in the YAML file is correct:
+		parsedServices := parseYAMLFileForNewTest(t, funcYAML)
+		services := *parsedServices
+
+		testProvider := stack.Provider{Name: "faas", GatewayURL: defaultGateway}
+		if !reflect.DeepEqual(services.Provider, testProvider) {
+			t.Fatalf("YAML `provider` section was not created correctly for file %s: got %v", funcYAML, services.Provider)
+		}
+
+		testFunction := stack.Function{Language: funcLang, Image: funcName, Handler: "./" + funcName}
+		if !reflect.DeepEqual(services.Functions[funcName], testFunction) {
+			t.Fatalf("YAML `functions` section was not created correctly for file %s, got %v", funcYAML, services.Functions[funcName])
+		}
+	}
+}
+
+func runAppendFunctionTest(t *testing.T, nft NewFunctionTest) {
+	defer func() {
```

Comment:
I think we should explicitly call this before every function rather than using defer.